src/sequence_batch_scheduler.cc

```diff
@@ -1586,12 +1742,12 @@ DirectSequenceBatch::BatcherThread(const int nice)
   }  // end runner loop
 
   LOG_VERBOSE(1) << "Stopping Direct sequence-batch scheduler thread "
-                 << batcher_idx_ << "...";
+                 << (size_t)model_instance_ << "...";
```

Comment:
Why not model_instance_->Name()? Like everywhere above?

Reply:
It is because the direct batcher thread is terminated after the model instances are destructed, which accessing the `Name()` could result in segmentation violation.
```
I0628 21:25:41.467416 38923 python_be.cc:2032] TRITONBACKEND_ModelFinalize: delete model state
I0628 21:25:41.467520 38923 sequence_batch_scheduler.cc:1105] Stopping sequence-batch reaper thread...
I0628 21:25:41.467596 38923 sequence_batch_scheduler.cc:1741] Stopping Direct sequence-batch scheduler thread <A1><9B>^Xb<A6>^?^@^@it_del_1...
I0628 21:25:41.467685 38923 sequence_batch_scheduler.cc:1741] Stopping Direct sequence-batch scheduler thread <FA><D9>.r<A4>^?^@^@it_del_0<A0>  ...
I0628 21:25:41.467734 38923 model_lifecycle.cc:602] OnDestroy callback() 'model_init_del' version 1
I0628 21:25:41.467746 38923 model_lifecycle.cc:604] successfully unloaded 'model_init_del' version 1
```

It looks like destructing the scheduler before any model instances solved the issue:
```
I0628 21:37:06.530392 42475 sequence_batch_scheduler.cc:1741] Stopping Direct sequence-batch scheduler thread model_init_del_0_2...
I0628 21:37:06.530561 42475 sequence_batch_scheduler.cc:1741] Stopping Direct sequence-batch scheduler thread model_init_del_0_1...
I0628 21:37:06.530638 42475 backend_model_instance.cc:828] Stopping backend thread for model_init_del_0_1...
I0628 21:37:06.530874 42475 python_be.cc:2193] TRITONBACKEND_ModelInstanceFinalize: delete instance state
I0628 21:37:07.637947 42475 backend_model_instance.cc:828] Stopping backend thread for model_init_del_0_2...
I0628 21:37:07.638081 42475 python_be.cc:2193] TRITONBACKEND_ModelInstanceFinalize: delete instance state
```
It is updated on this commit: [Destruct scheduler before instances on model destructor](https://github.com/triton-inference-server/core/pull/207/commits/10a0b7fd3fac8e5a1154b57625d2dfb66c72f312)

I wonder you could think of anything that might go wrong with the model destruction order change.

Reply:
We can avoid explicit thread destruction by storing the Name() value when model_instance is defined.

Reply:
I think the explicit destruction order is actually correct, because the destruction is changed from "[TritonModelInstance](https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L634-L637) -> Scheduler -> [TritonModel](https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L616)" to "[Scheduler](https://github.com/triton-inference-server/core/pull/207/commits/10a0b7fd3fac8e5a1154b57625d2dfb66c72f312#diff-1868cdcb6655ec791faf4b245a3ce92e4cd15983558118450208982edcc5fb5aR638) -> [TritonModelInstance](https://github.com/triton-inference-server/core/pull/207/commits/10a0b7fd3fac8e5a1154b57625d2dfb66c72f312#diff-1868cdcb6655ec791faf4b245a3ce92e4cd15983558118450208982edcc5fb5aR642-R644) -> [TritonModel](https://github.com/triton-inference-server/core/pull/207/commits/10a0b7fd3fac8e5a1154b57625d2dfb66c72f312#diff-1868cdcb6655ec791faf4b245a3ce92e4cd15983558118450208982edcc5fb5aR621)", which is exactly the opposite order of which they are constructed "[TritonModel](https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L165-L168) -> [TritonModelInstance](https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L244-L247) -> [Scheduler](https://github.com/triton-inference-server/core/blob/main/src/backend_model.cc#L249)".

We can also store the `Name()` value if you think updating the destruction order is too risky.

@GuanLuo @nnshah1 What do you think?

Reply:
The explicit destruction was removed with some design clean-ups if I remember correctly. I don't see any risks involved.